docs/determinism.md

```diff
@@ -0,0 +1,84 @@
+# Determinism 🤝 Rollups
+
+[Determinism](https://en.wikipedia.org/wiki/Deterministic_system#In_computer_science) is crucial for rollups because of [fault proof](https://community.optimism.io/docs/protocol/2-rollup-protocol/#fault-proofs). Validators will execute program([Turing machine](https://en.wikipedia.org/wiki/Turing_machine) tape) to validate the proposed state of rollup, syncing shared [repliced state machine](https://en.wikipedia.org/wiki/State_machine_replication). The state machine consists of a transition function. The transition function must be deterministic to decide which party is correct when dispute occur. For example, let's say that Alice is the state proposer and Bob is a validator. Bob will advance its own machine state by validating the program, and checks the sanity of proposed state at L1 and disagrees with it, and tells L1 that Alice has lied, triggering fault proof. If the program is nondeterministic, L1 will never find out who's claim(state) is right, leading L1 security inheritance loss.
```

Comment:
```suggestion
Rollups allow faster, cheaper, and more complex program executions because they execute programs off-chain and validate on on-chain. Thus, it is very important that all off-chain clients give the same output; otherwise, off-chain clients will incorrectly verify executions, cannot validate correctly on on-chain, and the rollup won't ever be able to be correctly verified. We call the validation process [fault-proof](https://community.optimism.io/docs/protocol/2-rollup-protocol/#fault-proofs), the characteristic that gives the expected output [Determinism](https://en.wikipedia.org/wiki/Deterministic_system#In_computer_science), and off-chain clients that validate the chain validator.

Let's take an Alice and Bob example where Alice is a state proposer, and Bob is a validator. Alice would propose a state at L1. However, when Bob advanced his own off-chain machine state machine, he received a different result. He triggered fault proof because he figured Alice lied. Here comes the importance of determinism. If the program is non-deterministic, L1 will never find out whose claim(state) is right, and the rollup no longer inherits L1's security.
```